### PR TITLE
Fix loaded data order bug when resuming from epoch >= 1

### PR DESCRIPTION
**Pull request by ngazagna-qc**
# What does this PR do?

Fixes #40690

Only 1 line added as described in the original issue.

Question: should we transform https://github.com/ngazagna-qc/transformers/blob/fix-data-order-resumed-epoch/reproduce_wrong_resumed_epoch.py into a test?

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation). -> Not required
- [ ] Did you write any new necessary tests? -> To discuss

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ngazagna-qc** [comment]
Hi @zach-huggingface @SunMarc : I am just following up on this bug and the proposed fix. It’s a one-line change with a reproducible example, and I’d love to help get it merged if possible. Let me know if you have any question :) 

Thanks a lot!

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2603,6 +2603,7 @@ def _inner_training_loop(
             steps_skipped = 0
             if steps_trained_in_current_epoch > 0:
                 epoch_dataloader = skip_first_batches(epoch_dataloader, steps_trained_in_current_epoch)
+                epoch_dataloader.iteration = epochs_trained
```
Comment (SunMarc) [inline src/transformers/trainer.py]: indeed this is probably linked to this also

**Review by SunMarc — COMMENTED** [review]
Thanks a lot ! Can you try to add some tests for this ? Maybe you can have a look at this [PR](https://github.com/huggingface/transformers/pull/40347/files) for the tests

File: src/transformers/trainer.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/trainer.py]: I think the simplest fix would be to move the following code just below the `if steps_trained_in_current_epoch > 0:` condition ! 
```python
            if hasattr(epoch_dataloader, "set_epoch"):
                epoch_dataloader.set_epoch(epoch)
```

**Comment by ngazagna-qc** [comment]
Hi @SunMarc , Thanks for your feedback. Sure, let me check the tests and try to implement one to spot and check the correction of the bug. I will probably work on it tomorrow.

File: src/transformers/trainer.py (hunk 1 above)
Comment (ngazagna-qc) [reply src/transformers/trainer.py]: > I think the simplest fix would be to move the following code just below the `if steps_trained_in_current_epoch > 0:` condition !
> 
> ```python
>             if hasattr(epoch_dataloader, "set_epoch"):
>                 epoch_dataloader.set_epoch(epoch)
> ```
I added the test and the fix suggested and we cannot **move** this snippet under the `if steps_trained_in_current_epoch > 0:` condition as the epoch needs to be set even if no batch has to be skipped. So instead I **copied** the snippet there :)

**Comment by ngazagna-qc** [comment]
@SunMarc, I think we are close to be able to merge this PR:
- I updated the trainer under the `if steps_trained_in_current_epoch > 0:` condition as we discussed
- I wrote the test in `tests/trainer/test_trainer.py::TrainerIntegrationTest::test_resume_batch_order`
- I confirmed that without the fix, the test fails, with the fix, the test passes. 

One can run `pytest -v -s tests/trainer/test_trainer.py::TrainerIntegrationTest::test_resume_batch_order` to see the output and understand better the issue.

I tried to follow the contribution guidelines but I am not super familiar with the commands to run to format the code. Let me know if anything more is needed :)

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2598,6 +2598,8 @@ def _inner_training_loop(
             if epoch == epochs_trained and resume_from_checkpoint is not None:
                 if steps_trained_in_current_epoch > 0 and not args.ignore_data_skip:
                     epoch_dataloader = skip_first_batches(epoch_dataloader, steps_trained_in_current_epoch)
+                    if hasattr(epoch_dataloader, "set_epoch"):
+                        epoch_dataloader.set_epoch(epoch)
```
Comment (SunMarc) [inline src/transformers/trainer.py]: delete the lines above then as we are duplicating the lines right now

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -5277,6 +5282,196 @@ def create_dummy_dataset():
         final_model_path = os.path.join(final_checkpoint_path, SAFE_WEIGHTS_NAME)
         self.assertTrue(os.path.exists(final_model_path), "Final model checkpoint was not saved!")
 
+    @require_safetensors
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]: seems a bit too complicated just for testing this, can you try to make this simpler ? Feel free to check the other tests

**Review by SunMarc — COMMENTED** [review]
Thanks ! Left a few comments

File: src/transformers/trainer.py (hunk 2 above)
Comment (ngazagna-qc) [reply src/transformers/trainer.py]: I do not think the lines above should be deleted. I ran tests with and without, and my test only passes if the `epoch_dataloader` sets the epoch in the main loop over epochs **AND** after `skip_first_batches` is called

File: tests/trainer/test_trainer.py (hunk 3 above)
Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py]: I have just simplified the test :) Sorry for the delay

Please let me know if you have other ideas of simplification.

**Comment by ngazagna-qc** [comment]
Hello @SunMarc , just a quick follow-up after 2 weeks :) I hope my updates are good enough now !

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -5197,6 +5202,179 @@ def create_dummy_dataset():
...
+        # 1 epoch consists of 10 points, so 5 steps with batch size 2
+        # Testing resuming from middle of epoch 1 (target_ckpt_num = 7)
+        import shutil
+
+        target_ckpt_num = 7
+        print(f"Deleting checkpoints from checkpoint-{target_ckpt_num} onwards...")
+        for ckpt_num in range(target_ckpt_num, last_ckpt_num + 1):
+            ckpt_path = os.path.join(exp_dir_checkpoint_deletion, f"checkpoint-{ckpt_num}")
+            if os.path.exists(ckpt_path):
+                shutil.rmtree(ckpt_path)
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]: you don't need to detele the checkpoints, just pass the correct checkpoint path.

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -5197,6 +5202,179 @@ def create_dummy_dataset():
...
+            seed=42,
+            learning_rate=0.1,
+            per_device_train_batch_size=2,
+            gradient_accumulation_steps=1,
+            save_strategy="steps",
+            save_steps=1,
+            num_train_epochs=3,
+            optim="sgd",
+            disable_tqdm=True,
+            dataloader_num_workers=0,  # Ensures that main process loads the data
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py]: why not resume from the baseline ? we are doing the same here

File: src/transformers/trainer.py (hunk 2 above)
Comment (SunMarc) [reply src/transformers/trainer.py]: hmmm that's strange, can you try to investigate a bit why this happens ? It's a bit counter-intuitive

**Review by SunMarc — COMMENTED** [review]
Thanks a lot ! Just some nits and we are good ot merge

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py]: Indeed, good catch !

File: tests/trainer/test_trainer.py (hunk 5 above)
Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py]: I just realized this too today. Indeed it's way easier! Doing it now and checking carefully there is no discrepancy in the behavior.

Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py]: done in [2d8b61b](https://github.com/huggingface/transformers/pull/40691/commits/2d8b61b68da7ecc9685b4655151f6bafd2e92d8d)

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py]: done in [2d8b61b](https://github.com/huggingface/transformers/pull/40691/commits/2d8b61b68da7ecc9685b4655151f6bafd2e92d8d)

File: src/transformers/trainer.py (hunk 2 above)
Comment (ngazagna-qc) [reply src/transformers/trainer.py]: I am exploring this. If I commented out/delete [these lines](https://github.com/ngazagna-qc/transformers/blob/2d8b61b68da7ecc9685b4655151f6bafd2e92d8d/src/transformers/trainer.py#L2397-L2398)
```
            if hasattr(epoch_dataloader, "set_epoch"):
                epoch_dataloader.set_epoch(epoch)
```
Then, for the baseline run, without resuming, I observe that the same randomness is applied at each epoch to sample batches:
```
Baseline: tensor([2, 6, 1, 8, 4, 5, 0, 9, 3, 7, 
                  2, 6, 1, 8, 4, 5, 0, 9, 3, 7, 
                  2, 6, 1, 8, 4, 5, 0, 9, 3, 7])
```
For comparison here is the **normal data order** (with current code base) **that my PR maintains**:
```
Baseline: tensor([2, 6, 1, 8, 4, 5, 0, 9, 3, 7, 
                  8, 4, 9, 0, 5, 1, 6, 7, 2, 3, 
                  2, 6, 3, 8, 7, 9, 1, 4, 5, 0])
```
I think this is because of [this line](https://github.com/ngazagna-qc/transformers/blob/2d8b61b68da7ecc9685b4655151f6bafd2e92d8d/src/transformers/trainer.py#L2396):
```
            epoch_dataloader = train_dataloader
```
that resets `epoch_dataloader` at every epoch and thus its `iteration` attribute to 0. 

I tried to put `epoch_dataloader = train_dataloader` before the for loop over epochs `for epoch in range(epochs_trained, num_train_epochs):` but I run into errors. I guess there was a reason to do so. 

**Conclusion:** we should then keep the update of the `iteration` attribute after `epoch_dataloader = train_dataloader`

**Comment by SunMarc** [comment]
make sure to rebase correctly, the diff is huge right now

**Comment by ngazagna-qc** [comment]
> make sure to rebase correctly, the diff is huge right now

I did reset from main. Let me know if anything else is due.

File: src/transformers/trainer.py (hunk 2 above)
Comment (ngazagna-qc) [reply src/transformers/trainer.py]: Does the above make sense @SunMarc ?

**Comment by ngazagna-qc** [comment]
Hi @SunMarc , I believe the rebase on main is correct now

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -5104,6 +5109,142 @@ def create_dummy_dataset():
...
+                data_indices = input_ids[:, 0].int()
+                self.data_order = torch.cat([self.data_order, data_indices.detach().clone()])
+
+                return {"loss": loss, "logits": logits}
+
+        # Scenario 1: Run baseline training to completion
+        # 1.1 Run training to completion
+        set_seed(42)
+        train_dataset = DummyDataset(size=10)
+        model_baseline = DummyModel(size=10)
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py:5174]: it only creates 3 checkpoints, I think the dataset you created is too small.

Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py:5174]: Checking this now

Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py:5174]: The training for 3 epochs over a dataset of 10 points, with a batch size of 2, no gradient accumulation leads correctly to the creation of 15 checkpoints. Indeed, an epoch requires 5 steps with bs=2.

I checked it :
```shell
$ pytest tests/trainer/test_trainer.py::TrainerIntegrationTest::test_resume_batch_order -v -s   
========================================================================================================================== test session starts ===========================================================================================================================
platform linux -- Python 3.10.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default'
rootdir: .../transformers
configfile: pyproject.toml
plugins: rerunfailures-15.1, rich-0.2.0, asyncio-1.2.0, xdist-3.8.0, order-1.3.0, hypothesis-6.140.3, anyio-4.11.0, timeout-2.4.0, cov-6.2.1
asyncio: mode=strict, debug=False, asyncio_default_fixture_loop_scope=function, asyncio_default_test_loop_scope=function
collected 1 item                                                                                                                                                                                                                                                         

tests/trainer/test_trainer.py::TrainerIntegrationTest::test_resume_batch_order {'train_runtime': 0.2858, 'train_samples_per_second': 104.973, 'train_steps_per_second': 52.486, 'train_loss': 2.2335431416829428, 'epoch': 3.0}
~~~~ Checking saved checkpoints after full training
checkpoint-13
checkpoint-10
checkpoint-14
checkpoint-15
checkpoint-4
checkpoint-11
checkpoint-6
checkpoint-9
checkpoint-5
checkpoint-3
checkpoint-8
checkpoint-2
checkpoint-7
checkpoint-1
checkpoint-12
{'train_runtime': 0.0479, 'train_samples_per_second': 626.658, 'train_steps_per_second': 313.329, 'train_loss': 1.27561403910319, 'epoch': 3.0}
~~~~ Checking saved checkpoints after resuming from ckpt 7
checkpoint-13
checkpoint-10
checkpoint-14
checkpoint-15
checkpoint-11
checkpoint-9
checkpoint-8
checkpoint-7
checkpoint-12
PASSED
```

Is there something I misunderstood? Or should I add comments if something is unclear?

Comment (SunMarc) [reply tests/trainer/test_trainer.py:5174]: hmmm i will try on my side again but I only saved 3 checkpoints. Can you try with transformers main ?

Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py:5174]: I rebased my code on the latests updates I could fetch from main yesterday. 

Let me do it again in a fresh env and a fresh branch directly checked out from current transformers main!

Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py:5174]: I just pull again the last main at this commit https://github.com/huggingface/transformers/commit/561233cabf480b0f884700be3aebf42dbe6a7b20 
I ran (inside my container), `pip uninstall transformers` and then `pip install -e ".[dev]"` (which output `Successfully installed transformers-5.0.0.dev0`) and the tests gives me the correct output as i shared above.

I do not know where this issue comes from. is it the same in the CI ?

Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py:5174]: Hi @SunMarc , Any update on this? I will try again from a fresh env tomorrow

**Comment by ngazagna-qc** [comment]
@SunMarc to reproduce:
```shell
git clone https://github.com/ngazagna-qc/transformers.git
cd transformers
uv venv .my-env
source .my-env/bin/activate
uv pip install -e ".[torch,sentencepiece,tokenizers,vision,integrations,timm,torch-vision,codecarbon,accelerate,video,num2words,mistral-common,chat_template,testing,quality,ja,sklearn,modelcreation]"
git checkout pr-fix-data-order-resumed-epoch
pytest tests/trainer/test_trainer.py::TrainerIntegrationTest::test_resume_batch_order -v -s
```

Let me know if I can help. I hope we are very close to merging.

**Comment by ngazagna-qc** [comment]
Hello, I just rebased to make it up to date @SunMarc and check if tests pass

File: tests/trainer/test_trainer.py [hunk 7]
```diff
@@ -5104,6 +5109,142 @@ def create_dummy_dataset():
         final_model_path = os.path.join(final_checkpoint_path, SAFE_WEIGHTS_NAME)
         self.assertTrue(os.path.exists(final_model_path), "Final model checkpoint was not saved!")
 
+    def test_resume_batch_order(self):
```
Comment (SunMarc) [inline tests/trainer/test_trainer.py:5120]:
```suggestion
    @require_torch_non_multi_accelerator
    def test_resume_batch_order(self):
```

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -2449,6 +2449,9 @@ def _inner_training_loop(
...
+
```
Comment (SunMarc) [inline src/transformers/trainer.py]: i checked that it should be fine to set it just one time just before `epoch_iterator = iter(epoch_dataloader)`. So we can remove this and the one at the start of the iteration

**Review by SunMarc — APPROVED** [review]
Thanks a lot, just checked that it was an issue on my side as i was running it with multiple gpu. Left some minor comments and we can merge !

File: tests/trainer/test_trainer.py (hunk 7 above)
Comment (ngazagna-qc) [reply tests/trainer/test_trainer.py:5120]: Done in [58600ac](https://github.com/huggingface/transformers/pull/40691/commits/58600acb7ba26618fcab2774bd190dfb0ede8d47)

File: src/transformers/trainer.py (hunk 8 above)
Comment (ngazagna-qc) [reply src/transformers/trainer.py]: Indeed, I checked it too. We now need only to set the epoch of the epoch dataloader only once.

Done in [fa179ab](https://github.com/huggingface/transformers/pull/40691/commits/fa179ab0e568f50b344f4b34fcd45ac7e951cd54)

**Comment by ngazagna-qc** [comment]
@SunMarc I think I am done with the updates :)
